### PR TITLE
feat(preview): attach to popup if already open

### PR DESCRIPTION
**Pull request by seblj**
- [X] Have you read [CONTRIBUTING.md](https://github.com/lewis6991/gitsigns.nvim/blob/HEAD/CONTRIBUTING.md)?

I agree with your comment here: https://github.com/lewis6991/gitsigns.nvim/pull/609#issuecomment-1243940330

Made it to always attach to the popup if it's already open

### CODE REVIEW

**Comment by seblj** [comment]
Maybe we should add an option `focusable: bool` or something in case someone opens a popup with an autocmd or something. I am not too sure if that is necessary. What do you think?

File: teal/gitsigns/popup.tl [hunk 1]
```diff
@@ -171,6 +171,15 @@ end
 local ns = api.nvim_create_namespace('gitsigns_popup')
 
 function popup.create(lines_spec: popup.LinesSpec, opts: {string:any}): integer, integer
+  -- Attach to window if already open
```
Comment (lewis6991) [inline teal/gitsigns/popup.tl:174]: This won't work if you open the preview and the blame sequentially, so we probably need to pass an id or some sort. Maybe we can repurpose the `gitsigns_preview` variable to be a string that is either `hunk` or `blame`.

We should also run this much earlier in actions.tl, before we create `lines_spec`.

Comment (seblj) [reply teal/gitsigns/popup.tl]: I added a window variable for diff and blame window and use this to fix this issue. I also try to attach to the popup earlier

**Comment by seblj** [comment]
I don't really like the name `popup.is_open0`, but I went with it because of `popup.create0`, and because I couldn't think of another name

File: teal/gitsigns/actions.tl [hunk 2]
```diff
@@ -645,7 +649,9 @@ M.preview_hunk = noautocmd(function()
     hunk = gs_hunks.patch_lines(hunk, vim.bo[bufnr].fileformat),
   })
 
-  popup.create(lines_spec, config.preview_config)
+  local winid, _ = popup.create(lines_spec, config.preview_config)
```
Comment (lewis6991) [inline teal/gitsigns/actions.tl:652]:
```suggestion
  local winid = popup.create(lines_spec, config.preview_config)
```

File: teal/gitsigns/actions.tl [hunk 3]
```diff
@@ -786,7 +797,9 @@ M.blame_line = void(function(opts: BlameOpts)
 
   scheduler()
 
-  popup.create(lines_format(blame_fmt, info), config.preview_config)
+  local winid, _ = popup.create(lines_format(blame_fmt, info), config.preview_config)
```
Comment (lewis6991) [inline teal/gitsigns/actions.tl:800]:
```suggestion
  local winid = popup.create(lines_format(blame_fmt, info), config.preview_config)
```

File: teal/gitsigns/popup.tl [hunk 4]
```diff
@@ -189,14 +189,28 @@ function popup.create(lines_spec: popup.LinesSpec, opts: {string:any}): integer,
   return winnr, bufnr
 end
 
-function popup.is_open(): boolean
+function popup.is_open0(var: string): boolean, integer
```
Comment (lewis6991) [inline teal/gitsigns/popup.tl:192]:
```suggestion
local function is_open(var: string): boolean, integer
```

File: teal/gitsigns/popup.tl [hunk 5]
```diff
@@ -189,14 +189,28 @@ function popup.create(lines_spec: popup.LinesSpec, opts: {string:any}): integer,
...
   for _, winid in ipairs(api.nvim_list_wins()) do
-    local exists = pcall(api.nvim_win_get_var, winid, 'gitsigns_preview')
+    local exists = pcall(api.nvim_win_get_var, winid, var)
```
Comment (lewis6991) [inline teal/gitsigns/popup.tl:194]:
```suggestion
    local exists = vim.w[winid][var]
```

File: teal/gitsigns/popup.tl [hunk 6]
```diff
@@ -189,14 +189,28 @@ function popup.create(lines_spec: popup.LinesSpec, opts: {string:any}): integer,
...
     if exists then
```
Comment (lewis6991) [inline teal/gitsigns/popup.tl:195]:
```suggestion
    if vim.w[winid][var] then
```

Comment (seblj) [reply teal/gitsigns/popup.tl:195]: Doing this requires to change `types.d.tl` so we don't get a compile error. However I couldn't get it to work. Any tips on how to do it?

**Comment by lewis6991** [comment]
I've gone ahead and fixed up the PR

**Comment by seblj** [comment]
> I've gone ahead and fixed up the PR

Thanks! That looks better

**Pull request merged**